src/System.Tuples/src/System/ValueTuple/ValueTuple.cs

```diff
@@ -0,0 +1,1119 @@
+        {
+            return ((IStructuralEquatable)this).GetHashCode(EqualityComparer<object>.Default);
+        }
+
+        int IStructuralEquatable.GetHashCode(IEqualityComparer comparer)
+        {
+            return ValueTuple.CombineHashCodes(comparer.GetHashCode(Item1), comparer.GetHashCode(Item2), comparer.GetHashCode(Item3));
+        }
+
+        int ITupleInternal.GetHashCode(IEqualityComparer comparer)
+        {
+            return ((IStructuralEquatable)this).GetHashCode(comparer);
+        }
+
+        public override string ToString()
+        {
+            return "(" + Item1 + ", " + Item2 + ", " + Item3 + ")";
+        }
+
+        int ITupleInternal.Size => 3;
+    }
+
+    public struct ValueTuple<T1, T2, T3, T4> : IEquatable<ValueTuple<T1, T2, T3, T4>>, IStructuralEquatable, IStructuralComparable, IComparable, ITupleInternal
+    {
+        public T1 Item1;
+        public T2 Item2;
+        public T3 Item3;
+        public T4 Item4;
+
+        public ValueTuple(T1 item1, T2 item2, T3 item3, T4 item4)
+        {
+            Item1 = item1;
+            Item2 = item2;
+            Item3 = item3;
+            Item4 = item4;
+        }
+
+        public override bool Equals(object obj)
+        {
+            return ((IStructuralEquatable)this).Equals(obj, EqualityComparer<object>.Default); ;
+        }
+
+        public bool Equals(ValueTuple<T1, T2, T3, T4> other)
+        {
+            return Equals(Item1, other.Item1) && Equals(Item2, other.Item2) && Equals(Item3, other.Item3) &&
+                Equals(Item4, other.Item4);
+        }
+
+        bool IStructuralEquatable.Equals(object other, IEqualityComparer comparer)
+        {
+            if (other == null || !(other is ValueTuple<T1, T2, T3, T4>)) return false;
+
+            var objTuple = (ValueTuple<T1, T2, T3, T4>)other;
+
+            return comparer.Equals(Item1, objTuple.Item1) && comparer.Equals(Item2, objTuple.Item2) && comparer.Equals(Item3, objTuple.Item3) &&
+                comparer.Equals(Item4, objTuple.Item4);
+        }
+
+        int IComparable.CompareTo(object obj)
+        {
+            return ((IStructuralComparable)this).CompareTo(obj, Comparer<object>.Default);
+        }
+
+        int IStructuralComparable.CompareTo(object other, IComparer comparer)
+        {
+            if (other == null) return 1;
+
+            if (!(other is ValueTuple<T1, T2, T3, T4>))
+            {
+                throw new ArgumentException(SR.ArgumentException_ValueTupleIncorrectType, nameof(other));
+            }
+
+            var objTuple = (ValueTuple<T1, T2, T3, T4>)other;
+
+            int c = comparer.Compare(Item1, objTuple.Item1);
+
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item2, objTuple.Item2);
+
+            if (c != 0) return c;
+
+            c = comparer.Compare(Item3, objTuple.Item3);
+
+            if (c != 0) return c;
+
+            return comparer.Compare(Item4, objTuple.Item4);
```

Comment:
Nit: I'd find the logic here a bit easier to follow if the spacing were a bit different:
```C#
int c = comparer.Compare(Item1, objTuple.Item1);
if (c != 0) return c;

c = comparer.Compare(Item2, objTuple.Item2);
if (c != 0) return c;

c = comparer.Compare(Item3, objTuple.Item3);
if (c != 0) return c;

return comparer.Compare(Item4, objTuple.Item4);
```
If you wanted, you could consolidate it even further would something like:
```C#
int c;
if ((c = comparer.Compare(Item1, objTuple.Item1) != 0) return c;
if ((c = comparer.Compare(Item2, objTuple.Item2) != 0) return c;
if ((c = comparer.Compare(Item3, objTuple.Item3) != 0) return c;
return comparer.Compare(Item4, objTuple.Item4);
```